sync/src/types.rs

```diff
@@ -945,15 +953,21 @@ impl SyncSnapshot {
     }
 
     pub fn get_ancestor(&self, base: &Byte32, number: BlockNumber) -> Option<core::HeaderView> {
-        // shortcut to return a ancestor block
-        if self.store().is_main_chain(&base) {
-            return self
-                .store()
-                .get_block_hash(number)
-                .and_then(|hash| self.get_header_view(&hash).map(HeaderView::into_inner));
-        }
-        self.get_header_view(base)?
-            .get_ancestor(number, |hash| self.get_header_view(hash))
+        let tip_number = self.tip_number();
+        self.get_header_view(base)?.get_ancestor(
+            number,
+            |hash| self.get_header_view(hash),
+            |current| {
+                // shortcut to return a ancestor block
+                if current.number() <= tip_number && self.store().is_main_chain(&current.hash()) {
```

Comment:
This may have a problem. If the chain forks between `if` and `get_block_hash`, this will return wrong result. We should use snapshot here.

Reply:
Done